### PR TITLE
Update to include JSON schema and some validation

### PR DESCRIPTION
**Pull request by scottdj92**

### CODE REVIEW

File: schema/valid.js [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+var json = require('./schema.json');
+
+module.exports = function validate(data) {
+	var validSchema = ajv.compile(json);
+	var valid = validSchema(data);
+
+	if(!valid) {
+		throw new Error("Your ExtractTextPlugin config is not correct. Please double check.");
+	} else {
+		console.log('your stuff works');
```
Comment (michael-ciniawsky) [inline schema/valid.js]: Is this printed on success every time ? 😛 I like the idea of a positive signal but maybe something like `PLUGIN_NAME:  ✅ `

Comment (bebraw) [inline schema/valid.js]: Yeah. Better just to be silent by default.

File: schema/schema.json [hunk 2]
```diff
@@ -0,0 +1,31 @@
+{
+  "$schema": "http://json-schema.org/draft-04/schema#",
+  "type": "object",
```
Comment (bebraw) [inline schema/schema.json]: You should probably set `"additionalProperties": false,` on root level as that gives stricter validation.

**Comment by d3viant0ne** [comment]
@scottdj92 - A couple of things before this can be landed.

1.) Need rebase to pick up some housekeeping in master
2.) It's currently failing on Travis, this will need to be addressed ( see review comment above )
3.) This could also use at least one test & it's negative counterpart

**Comment by bebraw** [comment]
Note that ajv can deal with default values. That could be another PR, though.

File: schema/valid.js [hunk 3]
```diff
@@ -8,7 +8,5 @@ module.exports = function validate(data) {
 
 	if(!valid) {
 		throw new Error("Your ExtractTextPlugin config is not correct. Please double check.");
```
Comment (bebraw) [inline schema/valid.js]: You can get a string message out of ajv. See https://github.com/webpack-contrib/purifycss-webpack for the basic idea. I do also defaults there.

**Comment by scottdj92** [comment]
Regarding adding tests, would it only pass testing if the value returned by ajv is `true`, meaning the passed in object matches the schema and vice versa? Or would the test have to go deeper than that?

**Comment by bebraw** [comment]
You can test schema success/failure at minimum, yeah.

**Comment by scottdj92** [comment]
Seeing [#1465](https://github.com/webpack/webpack/issues/1465) when running `npm test`. Not sure if webpack-core has been updated. Currently every other test passes, except for this one error.

File: schema/validator.js [hunk 4]
```diff
@@ -0,0 +1,12 @@
+var Ajv = require('ajv');
+var ajv = new Ajv({allErrors: true});
+var json = require('./schema.json');
+
+module.exports = function validate(data) {
+	var validSchema = ajv.compile(json);
+	var valid = validSchema(data);
+
+	if(!valid) {
+		throw new Error("Your ExtractTextPlugin config is not correct. Please double check.");
```
Comment (bebraw) [inline schema/validator.js]: You can get possible error text through `ajv.errorsText()`. Better throw that instead.

Comment (scottdj92) [inline schema/validator.js]: Added :D

**Comment by scottdj92** [comment]
All tests now pass as of 1/26/2017.
